### DISCUSSION TITLE
Scrapbook Application with Content/Template Types (text, video, image, etc) Schema

### DISCUSSION THREAD

**Discussion by jtkeyva** [supabase/supabase#7248] [Questions]
Wondering the best way to architect a scrapbook application that has different content types/templates: Text, Video, Image, Gif etc. A page can contain any number of content types ex:
 
A Collage Template can have 3 images, a video, and some text on it. 

Do I have a separate table for each type? How does the page know what table to look up? Does it have to scan each table?

Here's what I have but not sure how to piece it together. Seems like a lot of tables and a lot of lookups and redundancy, so do I have to lookup each table for items that match page_id? Would a join do this in one round trip? Would a jsonb column make sense? If I had say 20 different content types, this would be a lot to keep track of. Is there a more generic/universal way of doing it?

Thanks!

Table: pages

- id
- created_at
- created_by
- page_type
- hashtags

Table: text

- id
- text
- font
- dimensions
- coordinates
- color
- opacity
- rotation
- url

Table: video

- id
- page_id
- video_url
- dimensions
- coordinates
- opacity
- rotation
- duration
- transcription
- url

Table: image

- id
- page_id
- image_url
-  dimensions
- coordinates
- opacity
- rotation
- -url

Table: gif
-id
-page_id
-gif_url
- dimensions
- coordinates
- opacity
- rotation
- url

Table: audio
-id
-page_id
-audio_url
-duration
-transcription
-url

**Updated with more fields and changed from messaging to scrapbook for better explanation

**Comment by GaryAustin1** [comment]
Depending on your use of url's, it seems like you could just have the messages table with two extra columns, text and url.  Then based on the message_type, use the url appropriately or the text field.  Seems cleaner if the above are your only fields that are different and you don't need foreign keys to the videos/images/etc. in a table.  

If you start having things like length of video, type of image, size of image, etc. then it might make sense to have different tables.

**Reply by jtkeyva** [reply]
Thanks @GaryAustin1 I realized I wasn't clear so I updated my post. I forgot to mention that any template/page can have several pieces of content within so they are dynamic in nature.

To JSONB or not to JSONB?

Thanks!

**Reply by GaryAustin1** [reply]
Yeah, so then it is a harder decision with your added fields.  IF you need to search the details infrequently then jsonb is an idea to keep a single table.

I have a note taking app with 5 different kinds of notes.  I have a few key columns that are all shared for main searches and the jsonb for the note fields that can be specific to each note.  So one note table.  The nice thing is to add a new type of note, or new info to a type of note I don't have redo the table.  
The jsonb column lets me search for specifics when needed.  So normally just use the main columns to filter on types of notes for displaying latest notes.  Then for a detailed search use note type and user_id as part of the search (to limit data rows) then jsonb search for item specific to that note type.
Normally though, the jsonb is just loaded for display of note details.

**Reply by jtkeyva** [reply]
Yeah, I like the way your notes app sounds. 

For me, the canvas related images, video, text stuff: coordinates, dimensions, opacity & rotation pretty much won't ever need to be searched yet they share lots of the same data.

I do want to be able to search/filter the text field and the transcription fields but that's pretty much it. I will also want to filter by page_type and hashtags. Is searching/filtering JSONB not a great idea or difficult?

So you think this could be a solution?

Table:

pages

- id
- created_at
- created_by
- page_type
- tags
-content (JSONB)

Otherwise it seems like a bunch of joins etc.

Thanks!

**Comment by GaryAustin1** [comment]
I would not hesitate to have a field you search all the time have a copy in its own column (in addition to it being in the jsonb if it is reasonable size).  Empty columns when not used for a specific page type don't cost much at all. 

The joins are actually a bigger problem in this type of situation as they are in different tables and you can't have a column foreign key point to more than one table.  So you have to actually decide which table to read with a case statement or similar on your page_type and then use sql (or multiple client calls) to select that id from the specific table, and then put it together.

**Reply by jtkeyva** [reply]
Yes exactly, thanks. I was thinking that ya I could have the text I want to search in it's own column...but then if I had 3 text items in the template, do I concatenate and store all 3 text items in that dedicated column?

I hear you on JOINS, seems like it could take a toll and yes I want to avoid round trips to the db. I created a view that had 3 Left Joins...Is that overkill if I could just use JSONB?

Would GraphQL help me much in this situation?

**Comment by dshukertjr** [comment]
@jtkeyva 
Just passing through without any particular solution here, but I personally don't think you should be scared of performing joins. Postgres handles joins super well with proper indexes in place! Multiple left joins should not be too big of a deal at all!

BTW, [here](https://ocw.mit.edu/courses/1-264j-database-internet-and-systems-integration-technologies-fall-2013/resources/mit1_264jf13_lect_10/) are lecture slides at MIT about database normalizations and it is an amazing material to read when designing database in SQL. In general, you shouldn't go too wrong as long as you follow what is written here!

Also looping in our SQL expert here. @steve-chavez , do you have any good suggestion on how to structure the database in this particular situation?

**Reply by jtkeyva** [reply]
Hey thanks @dshukertjr good to know left joins aren't so bad. That was a good read! It does suggest having multiple people look data modeling so this discussion is right on par :)

Seems the canvas objects ex image, video, text, gif, sticker etc all share coordinates, dimensions, rotation and opacity. Perhaps there's a table called canvas_attributes or something? The only difference is that video has a duration (as does audio).

I look forward to @steve-chavez has to say.
